### DISCUSSION TITLE
Optimizing markup for texml and LaTeX

### DISCUSSION THREAD

**Discussion by davidmjones** [AmerMathSoc/texml#165] [Announcements]
Let's discuss how to handle cases where the optimal markup for LaTeX/print is not the optimal markup for texml.

To begin with, let's start listing the kinds of things you're running into, preferably with references to specific papers.

**Comment by djn-ams** [comment]
In my opinion, it is stylistically better to avoid breaking equations---even if they are disturbingly long---for MathViewer. My reasoning is threefold:[^1]

1.  It simply _looks_ better, to my eye. Also, we have infinitely scrolling text boxes, why not put them to good use? (Well... use.) Cf. [tran8916, Example 4.2](http://texml.ams.org/tran/376-07/tran8916/html/#ltxid101) for some displays that really did not need to be broken; they just look odd and out of place (and, in fact, this is also a good paper to illustrate my third point, and that alignments look better when used _properly_).
2. When breaking long displays, we may have to alter notation, which increases the chance for a notation mistake during editing. For example, think of when we are told to break a fraction with a denominator that spans the text width (or exponent).[^2]
3. Alignments, I think, carry implied meaning. Alignments are a good way of _semantically grouping_ related (or sequential) relations, computations, formulae, etc.

With that said, here are some of my observations.

-----

First, `multline` environments are _almost_ always used to get an equation to fit within the margins of a page; cf. [tran8760, Theorem 2.2](http://texml.ams.org/tran/376-04/tran8760/html/#ltxid13) for a particularly good example (there are two almost identical equations in the theorem; one is on one line and one is a `multline`, presumably because it did not fit on the page). Typically they are only used to break an wide equation at a relation, (binary) operator, or at a trailing quantifier. E.g.,
```
\begin{multline}
X\\
= Y {... very long equation...}
\end{multline}
```
or
```
\begin{multline}
X = Y + ... \\
+ {... more equation...}
\end{multline}
```
or 
```
\begin{multline}
X = Y + {... very long equation...}\\
\forall x\in Z
\end{multline}
```

Second, _certain_ alignments are used for nothing more than getting a __very__ long equation to fit on the page; again, see [tran8916, Example 4.2](http://texml.ams.org/tran/376-07/tran8916/html/#ltxid101). In cases such as
```
\begin{align}
X = {} & Y + ...\\
& {} + Z + ... \\
& {} + W + ...
\end{align}
```
we can be almost certain that the original equation was simply too wide to fit on the page and needed to be broken for the print product. Conversely, in cases such as
```
\begin{align}
X & = Y \\
& = Z \\
& = W 
\end{align}
```
we would _not_ want to reformat this as a single line equation as the alignment has implied meaning (these equalities follow from each other and are related) and, visually (and unfortunately, _only_ visually), it's easier to parse.

------

In light  of the first point, perhaps we could have TeXML automatically convert multlines to equations? Despite not being able to think of a good example, I could foresee this working flawlessly 98% of the time but the 2% of the time being enough of a problem that it would not be worth the trouble.

For the second point, I think the best way to go about it would be to add some syntactic sugar (perhaps to the `iftexml` package? The (internal) AMS document classes?) to provide an 'alignment, but only for print' environment, say `pralign`. Then we could have TeXML automatically convert this to an equation and disregard any alignment or row delineators (I say this as if I know it's easy, or even possible). This would have the added benefit that any edits to the TeX sources get made in _one_ place, rather than making multiple edits or having to fork the file. The one place I can see this going wrong would be if, during copy editing, we have something like so:
```
\begin{equation*}
X = Y ... ZW ...
\end{equation*}
```
and we are told to break it as:
```
\begin{align*}
X = {} & Y ... Z\\
& {} \times W ...
\end{align*}
```
If TeXML were to automatically flatten all alignment elements we'd end up with a spurious `\times` in our equation. Nonetheless, I think that this would result in better HTML, often enough, that it'd be worth the effort.

[^1]: As a fourth, unlisted benefit, it also "preserves the authors' vision/intent." Which we obviously care deeply, _deeply_ about.

[^2]: A good argument could be made that this is actually the _authors'_ fault(s) for formatting their result as poorly as possible. But unfortunately that's outside of our control at this point unless, as John Mulaney says, "everyone gets real cool about a bunch of stuff really quickly."

**Comment by pkra** [comment]
Some thoughts https://github.com/AmerMathSoc/texml/discussions/165#discussioncomment-6242157

I generally agree that it's better not to break up equations. However, for very different reasons and with different consequences.

The reason I favor to not unnecessarily break up long expression is that I hope we can develop means of enabling natural reflow in the resulting web-based outputs. Having hard breaks plays against that.

A good example are very large inline expressions. We currently have to break them up manually. However soon won't have to because we will get that automatically downstream.

Similarly, I hope that we can figure out ways to differentiate tabular layouts that are actually table-like (e.g., tables, matrices) from those used for alignment. Then we can consider downstream methods to reflow such content more naturally.

I don't think flattening tabular environments is a good way to get there however. Alignment information is very valuable for good reflow, too. 

I'd rather explore ways to match some tabular environments to reflow-optimized markup instead of just tex-math. 
A simple example might be our recent hacky "text" equations. We could similarly have a "simple" equation environment which is used for "very long, single-line expressions that look ok if they just wrap".

Finally, I want to comment one detail:

>  we have infinitely scrolling text boxes, why not put them to good use? 

While that's true for content delivered via the open web, this is not true for the more limited world of epub. 

Here, we cannot make things scrollable (because most epub reading systems remove/override the relevant CSS).

As a result, our only means is to shrink-to-fit equations. As a result, very wide expressions will become illegible on small viewports.  (And sadly most epub reading systems have poor zooming features.)
